From 8a0717cc07e78d1133cc10e62220770734181bee Mon Sep 17 00:00:00 2001
From: Tobias Deiminger <tobias.deiminger@linutronix.de>
Date: Fri, 30 Jan 2026 01:55:36 +0100
Subject: [PATCH] Check return value of fgets

The cmake project has a reasonable default setting
CMAKE_COMPILE_WARNING_AS_ERROR=ON in cmake/compiler_options.cmake. With this
setting our build will fail on unchecked return values from fgets(). This
patch fixes the issue by adding return value checks.

This is not just to silence the errors. C99 defines for fgets: "If a read
error occurs during the operation, the array contents are indeterminate and a
null pointer is returned". We should actually check the return value to avoid
undefined behavior from processing indeterminate data.

Let get_passcode_to_key_file return -1 on error, since it is used as callback
to OpenSSLs PEM_read_bio_PrivateKey_ex where documentation states "The
callback must return the number of characters in the passphrase or -1 if an
error occurred".

Upstream-Status: Pending
Origin: Debian [https://salsa.debian.org/collabora-team/imx-code-signing-tool/-/blob/fb2961efc88b9b6f37e98a5feed26873713299ee/debian/patches/check_return_values.patch]
---
 src/lib/back_end/pkey.c            |  6 +++++-
 src/tools/pki_tree/hab4_pki_tree.c | 25 +++++++++++++++++++++----
 2 files changed, 26 insertions(+), 5 deletions(-)

diff --git a/src/lib/back_end/pkey.c b/src/lib/back_end/pkey.c
index 31b5c3d..d6e9784 100644
--- a/src/lib/back_end/pkey.c
+++ b/src/lib/back_end/pkey.c
@@ -102,8 +102,12 @@ int get_passcode_to_key_file(char *buf, int size, int rwflag, void *userdata)
         return 0;
     }
 
-    fgets(buf, size, password_fp);
+    buf = fgets(buf, size, password_fp);
     fclose(password_fp);
+    if (buf == NULL)
+    {
+        return -1;
+    }
     chomp(buf);
 
     return strlen(buf);
diff --git a/src/tools/pki_tree/hab4_pki_tree.c b/src/tools/pki_tree/hab4_pki_tree.c
index 3c878e4..4ff15b2 100644
--- a/src/tools/pki_tree/hab4_pki_tree.c
+++ b/src/tools/pki_tree/hab4_pki_tree.c
@@ -118,6 +118,7 @@ int main(int argc, char **argv)
     int num_srk = 0;
     int val_period = 0;
     char *pass = NULL;
+    char *user_input = NULL;
     unsigned long serial = 0;
     char duration_str[10] = {0};
     char num_srk_str[10] = {0};
@@ -304,10 +305,18 @@ int main(int argc, char **argv)
         if (strcmp(existing_ca, "y") == 0)
         {
             printf("Enter CA key name: ");
-            fgets(ca_key, sizeof(ca_key), stdin);
+            user_input = fgets(ca_key, sizeof(ca_key), stdin);
+            if (user_input == NULL) {
+                fprintf(stderr, "Error while reading user input\n");
+                return 1;
+            }
             ca_key[strcspn(ca_key, "\n")] = '\0';
             printf("Enter CA certificate name: ");
-            fgets(ca_cert, sizeof(ca_cert), stdin);
+            user_input = fgets(ca_cert, sizeof(ca_cert), stdin);
+            if (user_input == NULL) {
+                fprintf(stderr, "Error while reading user input\n");
+                return 1;
+            }
             ca_cert[strcspn(ca_cert, "\n")] = '\0';
         }
         printf("\nKey type options (confirm targeted device supports desired "
@@ -330,10 +339,18 @@ int main(int argc, char **argv)
                             ARRAY_SIZE(rsa_kl_values));
         }
         printf("Enter PKI tree duration (years): ");
-        fgets(duration_str, sizeof(duration_str), stdin);
+        user_input = fgets(duration_str, sizeof(duration_str), stdin);
+        if (user_input == NULL) {
+            fprintf(stderr, "Error while reading user input\n");
+            return 1;
+        }
         duration = atoi(duration_str);
         printf("How many Super Root Keys should be generated ? : ");
-        fgets(num_srk_str, sizeof(num_srk_str), stdin);
+        user_input = fgets(num_srk_str, sizeof(num_srk_str), stdin);
+        if (user_input == NULL) {
+            fprintf(stderr, "Error while reading user input\n");
+            return 1;
+        }
         num_srk = atoi(num_srk_str);
         ask_until_valid(
             "Do you want the SRK certificates to have the CA flag set? (y/n) :",
-- 
2.47.3
